### PR TITLE
Merging the whole app

### PR DESCRIPTION
**Pull request by adammarostica**
- Added components for Welcome, CreatePoll, DisplayPoll, Voting, Results, and Result components.
- connected Firebase to CreatePoll, Voting, and Results components

### CODE REVIEW

File: src/App.js [hunk 1]
```diff
@@ -1,9 +1,34 @@
...
+  //   }
+  // }, [navigate])
+
   return (
-    <>
-    </>
+    <div className='app'>
+      <Header />
+      <main>
+        <Routes>
```
Comment (jhshim85) [inline src/App.js:23]: Love the idea of using Route!

File: src/CreatePoll.js [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+  
+  const [pollQuestion, setPollQuestion] = useState('');
+  const [pollOptions, setPollOptions] = useState([]);
+  const [slug, setSlug] = useState('');
+  const debouncedSlug = useDebounce(pollQuestion, 500);
+  const navigate = useNavigate();
+
+  // If each poll input has text, generate a new blank input
+  useEffect(() => {
+    if (pollOptions.every((option) => option.length > 0)) {
```
Comment (jhshim85) [inline src/CreatePoll.js:21]: great use of .every() method

File: src/CreatePoll.js [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+      console.log('nope');
+    }
+  }
+
+  return (
+    <section className="create">
+      <h2>Poll Creator</h2>
+      <form className="create__form" action="" onSubmit={handleSubmit}>
+        <label className="sr-only" htmlFor="poll-question">What's your question?</label>
+        <textarea onChange={handleQuestionChange} value={pollQuestion} name="poll-question" id="poll-question" cols="30" rows="2" placeholder="What's your question?" required></textarea>
```
Comment (jhshim85) [inline src/CreatePoll.js]: love the way it is, but may I suggest rows='3' instead?

File: src/CreatePoll.js [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+                  className="poll-option__input"
+                  onChange={handleOptionChange}
+                  onKeyDown={handleEnter}
+                  value={option}
+                  name={index}
+                  id={index}
+                  type="text"
+                  placeholder={
+                    index === 0
+                      ? "An answer"
```
Comment (jhshim85) [inline src/CreatePoll.js:163]: love how it automatically creates another answer input.
my question though, if there is no input, then the empty input doesn't get pushed into firebase, correct?

File: src/Results.js [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+    onValue(dbRef, (snapshot) => {
+      if (snapshot.exists()) {
+        const data = snapshot.val()
+        const newQuestion = data.question;
+        setPollQuestion(newQuestion);
+        const newResults = [];
+        for (let item in data.options) {
+          newResults.push([item, data.options[item]]);
+        }
+        newResults.sort((a, b)=> b[1] - a[1]);
```
Comment (jhshim85) [inline src/Results.js:26]: great use of sort() method

File: src/Results.js [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+      <h2>{pollQuestion}</h2>
+      {
+        pollResults.length > 0 ?
+          <section className="results__list">
+            {
+              pollResults.map((result) => (
+                <Result key={uuid()} resultValues={result} numTotalVotes={numTotalVotes} />
+              ))
+            }
+            <div className="results__list__percentage results__list__percentage--25">25%</div>
```
Comment (jhshim85) [inline src/Results.js:74]: may I suggest different BEM?
i.e. 'results__percentage--25'
'results__list--background'
'results__vote--count'

File: src/Voting.js [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+    if (e.key === 'Enter' || e.key === ' ') {
+      setUserChoice(e.target.innerText);
+    }
+  }
+
+  // Submit the user's vote and note that they have voted in both state and localStorage so that <DisplayPoll /> displays <Results /> instead.
+  function handleSubmit(e) {
+    e.preventDefault();
+    if (userHasVoted) {
+      alert('You have already voted in this poll.')
```
Comment (jhshim85) [inline src/Voting.js:48]: feel like prompt might be the better option

**Review by jhshim85 — COMMENTED** [review]
neat and great code overall
perfect use of routes and components

File: src/CreatePoll.js [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+
+export default function CreatePoll() {
+  
+  const [pollQuestion, setPollQuestion] = useState('');
+  const [pollOptions, setPollOptions] = useState([]);
+  const [slug, setSlug] = useState('');
+  const debouncedSlug = useDebounce(pollQuestion, 500);
+  const navigate = useNavigate();
+
+  // If each poll input has text, generate a new blank input
```
Comment (petrungaro) [inline src/CreatePoll.js:19]: This was very nice from a UI view, really smooth having the extra field pop up

File: src/CreatePoll.js [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+  useEffect(() => {
+    if (pollOptions.every((option) => option.length > 0)) {
+      const newOptions = [...pollOptions];
+      newOptions.push('');
+      setPollOptions(newOptions);
+    }
+  }, [pollOptions])
+
+  // When the user stops typing their question, check if a kebab-ified version of the question already exists as a Firebase property.
+  // If yes, the poll will be assigned a random code by Firebase. If no, prepare to set the poll with its custom code.
```
Comment (petrungaro) [inline src/CreatePoll.js:29]: Really nice work!

File: src/CreatePoll.js [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+    setPollQuestion(e.target.value);
+  }
+
+  function handleOptionChange(e) {
+    const newOptions = [...pollOptions];
+    newOptions[e.target.id] = e.target.value;
+    setPollOptions(newOptions);
+  }
+
+  // Prevent the enter key from submitting the form. Instead, make it focus on the next form element.
```
Comment (petrungaro) [inline src/CreatePoll.js:100]: Very nice! Love all the extra care you're taking to make sure its a good user experience

**Review by petrungaro — COMMENTED** [review]
Overall looks really really great, and I don't know enough about firebase to make any suggestions there. The error handling and edge cases you've covered really makes me want to go over my code again. Great job!

File: src/CreatePoll.js (hunk 3 above)
Comment (adammarostica) [reply src/CreatePoll.js]: Sounds good. Thanks!

File: src/CreatePoll.js (hunk 4 above)
Comment (adammarostica) [reply src/CreatePoll.js:163]: It won't push a poll without answers to firebase, but this comment has made me realize that I don't notify the user of what they need to do if they DO try to submit a poll without answers. The question is: do I add that functionality, or do I just leave people who try to submit answer-less polls to fend for themselves?

File: src/Voting.js (hunk 7 above)
Comment (adammarostica) [reply src/Voting.js:48]: Wouldn't prompt just ask them to provide text? I should mention that it should be impossible for anyone to ever see this text, as the Voting component should never render if the user has already voted.

**Pull request merged**